uncertainties/unumpy/core.py

```diff
@@ -30,10 +30,73 @@
     # Utilities:
     "nominal_values",
     "std_devs",
+    "average",
     # Classes:
     "matrix",
 ]
 
+
+def _average(arr):
+    """
+    The real implementation of average, for 1D arrays.
+    """
+    assert arr.ndim == 1
+    cov_matrix = covariance_matrix(arr)
+    weights = numpy.diagonal(cov_matrix) ** -1
+    weights_sum = weights.sum()
+    return uarray(
+        (nominal_values(arr) * weights).sum() / weights_sum,
+        numpy.sqrt(
+            numpy.einsum(
+                "i,ij,j",
+                weights,
+                cov_matrix,
+                weights,
+            )
+        ) / weights_sum,
+    )
+
+def average(arr, axes=None):
+    """
+    Return a weighted averaged along with a weighted mean over a certain axis
+    or a axes. The formulas implemented by this are:
+
+    $$ \mu = \frac{\sum_i (x_i/\sigma_i^2)}{\sum_i \sigma_i^{-2}}$$
+
+    $$\sigma_\mu = \frac{\sqrt{\sum_{i,j} \sigma_i^{-2} \sigma_j^{-2} \cdot Cov(x_i, x_j)}}{\sum_i \sigma_i^{-2}}$$
+
+    Where of course $Cov(x_i, x_i) = \sigma_i^2$.
+
+    By default, operates on all axes of the given array.
+    """
+    arr = np.asanyarray(arr)
+    if axes is None:
+        axes = [0]
+        arr = arr.flatten()
+    # The following sanity checks on axes similar to what Numpy 2.x's
+    # lib.array_utils.normalize_axis_tuple do.
+    elif not isinstance(axes, [tuple, list]):
+        axes = [operator.index(axes)]
+    else:
+        axes = tuple(axes)
+    if len(set(axes)) != len(axes):
+        raise ValueError("repeated axis found in axes argument")
+    # This one is not checked by np.lib.array_utils.normalize_axis_tuple
+    if max(axes) >= arr.ndim:
+        raise ValueError(
+            f"Cannot average over an inexistent axis {max(axes)} >= arr.ndim = "
+            f"{arr.ndim}"
+        )
+    if not isinstance(arr.flat[0], core.Variable):
+        raise ValueError(
+            "unumpy.average is meant to operate upon numpy arrays of ufloats, "
+            "not pure numpy arrays"
+        )
+    for axis in sorted(axes, reverse=True):
+        arr = numpy.apply_over_axis(_average, axis, arr)
```

Comment:
Hmm on a 2nd thought, perhaps this means that e.g for `A` with `shape=(3,4,5,7)`, and `axes=(1,3)`, if there are correlations between `A[:,0,:,0]` and `A[:,1,:,1]` are not taken into account? Because once `axis=1` is averaged, the `covariance_matrix` call for the 1d slices of axis 1 won't take into account the correlations to the values already averaged on axis 3...

I think we can live with that, but perhaps warn the users about it in the function's __doc__, or elsewhere. Unless of course someone here will think of a better way to implement this.